client/src/components/Login.js

```diff
@@ -0,0 +1,311 @@
+import registeredMethodType from 'types/registeredMethod';
+
+class Login extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      selectedMethod: null,
+      loginProps: null,
+      message: null,
+      showOtherMethods: false,
+    };
+
+    this.handleCompleteLogin = this.handleCompleteLogin.bind(this);
+    this.handleShowOtherMethodsPane = this.handleShowOtherMethodsPane.bind(this);
+    this.handleHideOtherMethodsPane = this.handleHideOtherMethodsPane.bind(this);
+    this.handleClickOtherMethod = this.handleClickOtherMethod.bind(this);
+  }
+
+  componentDidMount() {
+    const { defaultMethod, registeredMethods } = this.props;
+
+    // Choose either the default method or the first method in the list as the default login screen
+    const defaultMethodDefinition = defaultMethod && registeredMethods.find(
+      method => method.urlSegment === defaultMethod
+    );
+
+    if (defaultMethodDefinition) {
+      this.setSelectedMethod(defaultMethodDefinition);
+    } else {
+      // TODO is this expected? We have the "first" method as the fallback default?
+      this.setSelectedMethod(registeredMethods[0]);
+    }
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    const { selectedMethod } = this.state;
+
+    // If the selected method has changed (or been set for the first time) then we'll load a "start"
+    // endpoint to get the process going
+    if (
+      (!prevState.selectedMethod && selectedMethod)
+      || (prevState.selectedMethod
+        && selectedMethod
+        && prevState.selectedMethod.urlSegment !== selectedMethod.urlSegment
+      )
+    ) {
+      this.fetchStartLoginData();
+    }
+  }
+
+  /**
+   * Set the current method the user will use to complete authentication
+   *
+   * @param {Object} method
+   */
+  setSelectedMethod(method) {
+    this.setState({
+      selectedMethod: method,
+      // When a method is chosen we'll assume the "select other method" screen is not relevant now
+      showOtherMethods: false,
+    });
+  }
+
+  /**
+   * Helper function to return methods aside from the selected one
+   *
+   * @return {Object[]}
+   */
+  getOtherMethods() {
+    const { registeredMethods } = this.props;
+    const { selectedMethod } = this.state;
+
+    if (!selectedMethod) {
+      return registeredMethods;
+    }
+
+    return registeredMethods.filter(method => method.urlSegment !== selectedMethod.urlSegment);
+  }
+
+  /**
+   * Trigger a "fetch" of state for starting a login flow
+   */
+  fetchStartLoginData() {
+    const { endpoints: { login } } = this.props;
+    const { selectedMethod } = this.state;
+
+    const endpoint = login.replace('{urlSegment}', selectedMethod.urlSegment);
+
+    this.setState({
+      loading: true,
+    });
+
+    // "start" a login
+    fetch(endpoint).then(response => response.json().then(result => {
+      this.setState(() => ({
+        loading: false,
+        loginProps: result,
+      }));
```

Comment:
Is it necessary to return a callback here rather than an object? Is this because this is inside a promise callback itself?

Reply:
You're right it's unnecessary. It would be if it relied on existing state for the update because it's in a promise callback but it doesn't rely on existing state. I'll change it back.